tests/integration/test_client_http.py

```diff
@@ -0,0 +1,203 @@
+import time
+import select
+import socket
+import random
+import threading
+from tests import unittest
+from contextlib import contextmanager
+
+import botocore.session
+from botocore.config import Config
+from botocore.vendored.six.moves import BaseHTTPServer, socketserver
+from botocore.exceptions import (
+    ConnectTimeoutError, ReadTimeoutError, EndpointConnectionError,
+    ConnectionClosedError,
+)
+
+
+class TestClientHTTPBehavior(unittest.TestCase):
+    def setUp(self):
+        self.port = random_port()
+        self.localhost = 'http://localhost:%s/' % self.port
+        self.session = botocore.session.get_session()
+
+    def test_can_proxy_https_request_with_auth(self):
+        proxy_url = 'http://user:pass@localhost:%s/' % self.port
+        config = Config(proxies={'https': proxy_url})
+        client = self.session.create_client('ec2', config=config)
+
+        class AuthProxyHandler(ProxyHandler):
+            def validate_auth(self):
+                proxy_auth = self.headers.get('Proxy-Authorization')
+                return proxy_auth == 'Basic dXNlcjpwYXNz'
+
+        try:
+            with background(run_server, args=(AuthProxyHandler, self.port)):
+                client.describe_regions()
+        except BackgroundTaskFailed:
+            self.fail('Background task did not exit, proxy was not used.')
+
+    def test_read_timeout_exception(self):
+        config = Config(read_timeout=0.1, retries={'max_attempts': 0})
+        client = self.session.create_client('ec2', endpoint_url=self.localhost,
+                                            config=config)
+
+        class FakeEC2(SimpleHandler):
+            msg = b'<response/>'
+            def get_length(self):
+                return len(self.msg)
+
+            def get_body(self):
+                time.sleep(1)
+                return self.msg
+
+        try:
+            with background(run_server, args=(FakeEC2, self.port)):
+                client.describe_regions()
+        except ReadTimeoutError:
+            # Note: This was vendored requests ConnectionError & ReadTimeout
+            pass
+        except BackgroundTaskFailed:
+            self.fail('Fake EC2 service was not called.')
+        else:
+            self.fail('Excepted exception was not thrown')
+
+    def test_connect_timeout_exception(self):
+        config = Config(connect_timeout=0.2, retries={'max_attempts': 0})
+        client = self.session.create_client('ec2', endpoint_url=self.localhost,
+                                            config=config)
+
+        def no_accept_server():
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            sock.bind(('', self.port))
+            time.sleep(4)
```

Comment:
Can we use `threading.Events` instead?  We had similar sleeps in chalice and it ended up being flaky.  I think threading events will make these tests more robust.

Reply:
Yeah, this sounds good. Done.